plugins/trace_api_plugin/compressed_file.cpp

```diff
@@ -0,0 +1,308 @@
+
+#include <zlib.h>
+
+namespace {
+   using seek_point_entry = std::tuple<uint64_t, uint64_t>;
+}
+
+namespace eosio::trace_api {
+
+struct compressed_file_impl {
+   static constexpr size_t read_buffer_size = 4*1024;
+   static constexpr size_t compressed_buffer_size = 4*1024;
+
+   ~compressed_file_impl()
+   {
+      if (initialized) {
+         inflateEnd(&strm);
+         initialized = false;
+      }
+   }
+
+   void read( char* d, size_t n, const std::unique_ptr<fc::cfile>& file_ptr )
+   {
+      if (!initialized) {
+         if (Z_OK != inflateInit2(&strm, -15)) {
+            throw std::runtime_error("failed to initialize decompression");
+         }
+
+         remaining_read_buffer = 0;
+         strm.avail_in = 0;
+         initialized = true;
+      }
+
+      size_t written = 0;
+
+      // consume the left over from the last read if there is any
+      if (remaining_read_buffer > 0) {
+         auto to_read = std::min(remaining_read_buffer, n);
+         std::memcpy(d, read_buffer.data(), to_read );
+         remaining_read_buffer -= to_read;
+         written += to_read;
+
+         if ( remaining_read_buffer > 0 ) {
+            std::memmove(read_buffer.data(), read_buffer.data() + to_read, remaining_read_buffer);
+            return;
+         }
+      }
+
+
+      // decompress more chunks
+      while (written < n) {
+         if ( strm.avail_in == 0 ) {
+            size_t remaining = file_size - file_ptr->tellp();
+            size_t to_read = std::min((size_t)compressed_buffer.size(), remaining);
+            file_ptr->read(reinterpret_cast<char*>(compressed_buffer.data()), to_read);
+            strm.avail_in = to_read;
+            strm.next_in = compressed_buffer.data();
+         }
+
+         do {
+            strm.avail_out = read_buffer.size();
+            strm.next_out = read_buffer.data();
+            auto ret = inflate(&strm, Z_NO_FLUSH);
+
+            if (ret == Z_NEED_DICT || ret == Z_DATA_ERROR || ret == Z_MEM_ERROR) {
+               throw std::runtime_error("Error decompressing: " + std::string(strm.msg));
+            }
+
+            auto bytes_decompressed = read_buffer.size() - strm.avail_out;
+            auto to_copy = std::min(bytes_decompressed, n - written);
+            std::memcpy(d + written, read_buffer.data(), to_copy);
+            written += to_copy;
+
+            if (bytes_decompressed > to_copy) {
+               // move remaining to the front of the buffer
+               std::memmove(read_buffer.data(), read_buffer.data() + to_copy, bytes_decompressed - to_copy);
+               remaining_read_buffer = bytes_decompressed - to_copy;
+            }
+
+            if (written < n && ret == Z_STREAM_END) {
+               throw std::runtime_error("Attempting to read past the end of a compressed file");
+            }
+         } while (strm.avail_out == 0 && written < n);
+      }
+   }
+
+   void seek( long loc, const std::unique_ptr<fc::cfile>& file_ptr  ) {
```

Comment:
Minor thing, but I would make this take a `fc::cfile&` instead of an `unique_ptr` and `const` here is a bit misleading.